Add load balancer info (#13)
mulesoft/docs-govcloud: modules/ROOT/pages/gov-cloud-load-balancer.adoc

= Dedicated Load Balancer Architecture
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> A Government Cloud dedicated load balancer enables you to route external HTTP and HTTPS traffic to multiple 

Reviewer: ```suggestion
A MuleSoft Government Cloud dedicated load balancer enables you to route external HTTP and HTTPS traffic to multiple 
```

> Mule applications deployed into Government Cloud workers inside a Virtual Private Cloud (VPC). Each Government Cloud

Reviewer: ```suggestion
Mule applications deployed to MuleSoft Government Cloud workers inside Anypoint Virtual Private Cloud (Anypoint VPC). Each MuleSoft Government Cloud
```

> dedicated load balancer is assigned to an Anypoint VPC, and that Anypoint VPC runs within a particular service region. 

Reviewer: ```suggestion
dedicated load balancer is assigned to an instance of Anypoint VPC, and that instance of Anypoint VPC runs within a particular service region. 
```

> The dedicated load balancer routes both external and VPC-internal traffic to Government Cloud workers deployed into that 

Reviewer: ```suggestion
The dedicated load balancer routes both external and Anypoint VPC-internal traffic to MuleSoft Government Cloud workers deployed to that instance of
```

Anypoint VPC.

> 

Reviewer: Do you suppose we should include one or more illustrations for the previous paragraphs, since this IS an architecture topic?

Reviewer: GLOBALLY search for and correct product naming to _MuleSoft Government Cloud_ and _Anypoint VPC_.
Do not use _Government Cloud_ or _VPC_.
I will stop marking this, except where I am suggesting other changes.

> The Government Cloud load balancer already redirects these requests, where `_myApp_` is the name of the Mule 

Reviewer: What do we mean by "already"? As opposed to what we just said above it? Aren't they the same thing?

application deployment to Government Cloud:

* HTTP requests to `+http://myApp.cloudhub.io+` redirects to `mule-worker-myApp.cloudhub.io:8081`
* HTTPS traffic to `+https://myApp.cloudhub.io:443+` redirects to `+https://mule-worker-myApp.cloudhub.io:8082+`

> Your load balancer listens for outside requests over HTTPS and, by default, communicates internally with your 

Reviewer: Now it's "your" load balancer? Do we mean the MGC load balancer used for your Mule application?
Can we use personalized narrative for the beginning of this topic as well: "you" and "your"?

worker over HTTP. If you configured your Mule application within the Anypoint VPC to listen on HTTPS, make sure 
you set `upstreamProtocol` to `HTTPS` when creating the mapping list using the 

> xref:anypoint-platform-cli-commands.adoc#cloudhub-load-balancer-mappings-add[`load-balancer mappings add`] command.

Reviewer: This looks like something that should be part of the TASK of configuring your app withing AVPC to listen on HTTPS, rather than here. "Making sure" is not something the audience should have to do in this intro context.

Reviewer: Also, if it is the **default** behavior for your load balancer to listen over HTTPS, why would the audience have to set this?

Reviewer: Just a general comment: it would nice to prepare the audience for the subtopics by listing them in order and using the same terms as the subtopic titles, something like...:

The MuleSoft Government Cloud dedicated load balancer is designed to:

* Accept external requests
* Rename external Mule application URLs
* Route internal requests
* Pass or decode encoded requests, as you choose

> == Routing External Requests to the Dedicated Load Balancer

Reviewer: GLOBALLY reword to avoid beginning a topic title with a gerund, which implies a Task topic (for which we use imperative voice).

Reviewer: How is this a subtopic of dedicated load balancer architecture?

Reviewer: ```suggestion
== How External Requests Are Routed to the Dedicated Load Balancer
```
?

A Government Cloud dedicated load balancer provides an alternative domain name to route HTTP requests to 
Mule applications listening on port 8091 and HTTPS requests to Mule applications listening on port 8092. You can 
also use mapping rules to rename requests to the Government Cloud dedicated load balancer to a different Mule 

> application domain name.

Reviewer: Is the second sentence a different method of doing the same thing: providing an alternate domain name?
Clarify: the GC provides a name, or you can manually map to a name?

> The Government Cloud dedicated load balancer exposes an external domain name that resolves to two public IP 

Reviewer: "exposes" = "provides an alternative" to "two public IP addresses" = 8091 and 8092?
If so, why repeat?
Avoid restatement (redundant), especially if using different wording to mean the same thing.

addresses which are accessible from outside your Government Cloud Anypoint VPC network. `_lb-name_` is the name you 
gave the load balancer when you created it:

> 

Reviewer: So, the external domain name is `_lb-name_`? Why are we showing an example of it; what does this example illustrate with regard to the previous narrative?

----
 _lb-name.lb-gprod-rt.anypointdns.net
----

> == Renaming External Mule Application URLs Through a Dedicated Load Balancer

Reviewer: Is this a user task? 
If so, provide steps and retitle to use imperative voice.
If not, retitle to not use gerund case.

Reviewer: ```suggestion
== How to Use a Mapping Rule to Rename External Mule Application URLs 
```
?

Reviewer: Also...how is this a subtopic of architecture?

> You can create a mapping rule to rename the Mule application, so it can be accessed through a different domain name.

Reviewer: ```suggestion
You can create a mapping rule to rename your Mule application so that <blah> can access it. 
```

For example, suppose you deploy a Mule application named myApp-myCompany-prod.cloudhub.io. Government Cloud domain names 
must be globally unique among every other MuleSoft Government Cloud customer and Mule application. A dedicated load balancer helps 

> you hide this naming complexity within your corporate DNS domain name.

Reviewer: This text should not follow the example name, because it explains something that is not made clear yet. I think we want to say that the name must be unique in gov cloud, and that your DLB helps you to meet this requirement...for example, take <this name>. You can make it unique by <doing this>.

Reviewer: In the interest of meeting your deadline, I am abandoning org. and structural editing at this point but continuing with simple copyediting. However, I think you can use the comments provided thus far to work on targeting the stated topic of Architecture or change the focus to user tasks, and make the narrative more clearly follow a (chrono)logical path.

You can set a mapping rule in your dedicated load balancer so external clients can access the Mule application 

> on `+http://myApp.lb-name.lb-gprod-rt.anypointdns.net+`, or `+https://myApp.lb-name.lb-gprod-rt.anypointdns.net:443+`.

Reviewer: ```suggestion
on either `+http://myApp.lb-name.lb-gprod-rt.anypointdns.net+` or `+https://myApp.lb-name.lb-gprod-rt.anypointdns.net:443+`.
```

> In addition, you can set a CNAME record in your corporate DNS nameserver to mask the `lb-name.lb-gprod-rt.anypointdns.net` 
> domain to your company's "vanity" domain.

Reviewer: Will your audience be familiar with CNAME?

```suggestion
In addition, you can set a CNAME record in your corporate DNS nameserver to mask the `lb-name.lb-gprod-rt.anypointdns.net` domain by using instead your company's "vanity" domain.
```
And will they know what a "vanity" domain is?

> For example, suppose your company owns the DNS domain `example.com`. You can create a CNAME record to route requests 

Reviewer: ```suggestion
For example, suppose your company owns the DNS "vanity" domain `example.com`. You can create a CNAME record to mask requests 
```

> to `myapp.example.com` to `myApp.lb-name.lb-gprod-rt.anypointdns.net`. This hides the complexity of the Government Cloud domain name 

Reviewer: ```suggestion
to `myApp.lb-name.lb-gprod-rt.anypointdns.net` by routing them instead to `myapp.example.com`, thereby hiding the complexity of the MuleSoft Government Cloud domain name 
```

> in the `cloudhub.io` domain.

Reviewer: Do we need this bit? this domain is not intro'd above.

== Routing Internal Requests to the Dedicated Load Balancer

> The Government Cloud dedicated load balancer has an internal domain name that is used by applications and clients 

Reviewer: an internal domain name or just an internal domain?
```suggestion
MuleSoft Government Cloud dedicated load balancer has an internal domain for applications and clients 
```

> within the Anypoint VPC.

Reviewer: ```suggestion
within Anypoint VPC.
```

> The internal domain uses the following naming convention (`_lb-name_` is  the name you gave the load balancer 

Reviewer: ```suggestion
The internal domain is named using the following convention, in which `_lb-name_` is  the name you gave your load balancer when you created it:
```

> when you created it):

Reviewer: See above.

Using "+" at the beginning of a line in a paragraph breaks the paragraph so that it can be edited only by the portion established by the +. It is much easier for me to edit (and you to understand edits) if you do not use these within paragraphs.

----
internal-_lb-name_.lb-gprod-rt.anypointdns.net
----

> Each Dedicated Load Balancer has a DNS A record `_lb-name_.lb-gprod-rt.anypointdns.net` that resolves to the two public 

Reviewer: Why are we repeating this, which is the first sentence of this subtopic, but WITHOUT the further clarification of the name being "a DNS A record". Clarify intro, and then shorten this:
```suggestion
This name resolves to the two public 
```

> IP addresses of the two instances. Through your DNS provider, you can add a CNAME record pointing to this A record 

Reviewer: What are the two IP addresses and two instances? 
I believe this text is in the intro to the topic set; suggest moving it HERE, where it is fully described.

and use your domain names to access it.

If you want your load balancer to manage all connections to your application but do not want your default domain 
name for your application publicly exposed, then each application must listen on HTTP port 8091 or 8092. Alternatively, 

> you can create a custom mapping policy to redirect outside requests from your load balancer to your specific application.

Reviewer: Again, wasn't all this covered already?
Keep topic set intros brief by mentioning but not describing the subtopics in the intro to the set.
A set intro should cover:
* What this topic set is about
* Why your audience should care 
* Summary of subtopics
GLOBAL

> The following graphic shows how a load balancer interacts between a VPC and Anypoint Platform.

Reviewer: ```suggestion
The following graphic shows how a load balancer interacts with an instance of Anypoint VPC and Anypoint Platform:
```

image::pxty-lb.png[Load Balancer]

> Your load balancer listens for outside requests over HTTPS and, by default, communicates internally with your worker 

Reviewer: It is still unclear to me, representing the new user, how 
load balancer + outside requests + HTTPS + worker + HTTP 
relates to 
Mule app + VPC + HTTPS + upstream protocol + mapping

over HTTP. If you configured your Mule application within the Anypoint VPC to listen on HTTPS, you must 
set `upstreamProtocol` to HTTPS when creating the mapping list using the `load-balancer mappings add` command.

> The internal HTTP mode lets you configure how HTTP requests are managed. They are either silently ignored, 

Reviewer: ```suggestion
The internal HTTP mode enables you to configure how HTTP requests are managed. They are either silently ignored, 
```
Does "either" refer to two options or are there really three here?
If three, consider a bullet list.

> mapped HTTPS on the default SSL endpoint, or the URL is mapped straight through to the corresponding HTTPS URL.

Reviewer: ```suggestion
mapped to HTTPS on the default SSL endpoint, or the URL of the incoming HTTP request is mapped directly to the corresponding HTTPS URL.
```

> == DLB URL Encoding

Reviewer: ```suggestion
== Dedicated Load Balancer URL Encoding
```

> 

Reviewer: Does this subtopic belong within one of the previous topics? It seems strangely isolated.

> You can configure the DLB to pass request URIs unchanged or decoded:

Reviewer: GLOBALLY replace "DLB" with spelling.

Reviewer: Clarify the choice between using URI vs. URL; which one are we really documenting here?

> * If you select *Keep URL encoding*, the DLB passes only the %20 and %23 characters as is.

Reviewer: If it retains only two characters, how is that "keeping" the encoding?
What happens to the other characters?
